app/models/config.rb

```diff
@@ -70,4 +95,100 @@ def self.start_day
     start ||= "monday"
     start.downcase.to_sym
   end
+
+  private
+    ### Pre-validation cleanup
+
+    def clean_config_value
+      # do nothing if empty
+      return if config_key.nil? || options.last.nil?
+
+      clean_f = CLEAN_PRE_VALIDATION[config_key.to_sym]
+
+      # no clean function, return
+      return if clean_f.nil?
+
+      method(clean_f).()
+    end
+
+    def clean_url
+      url = options.last
+
+      # don't try to clean empty url
+      return if url.blank?
+
+      # don't have to do anything, already https
+      return if url.start_with? 'https://'
+
+      # convert http or // to https://
+      if url.start_with? /(http:)?\/\//
+          url = url.sub /^(http:)?\/\//, 'https://'
+
+      # convert no scheme to https:// (i.e. example.com -> https://example.com)
+      elsif not url.start_with? '/'
+          url = 'https://' + url
+      end
+
+      # set config back to what it was
+      config_value['options'] = [url]
+    end
+
+    def fix_capitalization
+      config_value['options'] = [options.last.capitalize]
+    end
+
+    ### Validation
+
+    # parent function. will handle errors; child validators should return true
+    # if value is valid for key, and false otherwise.
+    def validate_config
+      # don't try to validate if no key or no value
+      return if config_key.nil? || options.last.nil?
```

Comment:
Two things:
* when is `config_key` null? Is that something we have to look out for or are you playing it safe? If the latter, would love to take it out and just do `options.last.blank?`
* let's use `.blank?` for consistency - it's a bit more aggro about whitespace which is good here

Reply:
IIRC I added this because some of the more pathological tests attempt to pass in a `nil` key. From what I understand of the system this would never happen in real life unless you, like, were messing around with developer tools and manually sending POST requests.